Sync/BatchingClient.swift

```diff
@@ -0,0 +1,221 @@
+        }
+
+        // If we can fit in another payload, push it up using our batch token
+        if batchMeta.add(self.currentPayload) {
+            return self.uploadPayload(self.currentPayload, queryParams: [NSURLQueryItem(name: "batch", value: token)])
+                >>> effect({ self.currentPayload = Payload(config: self.config) })
+                >>> succeed
+        }
+
+        // Otherwise, 
+        // 1. Commit the batch
+        // 2. Start up a new batch
+        // 3. Place the payload into the new batch
+        return self.commit(token) >>> {
+            return self.start(self.currentPayload)
+                >>== effect({ self.batchToken = $0 })
+                >>> effect({ self.batchMeta = BatchMeta(config: self.config) })
+                >>> effect({ self.batchMeta.add(self.currentPayload) })
+                >>> succeed
+        }
+    }
+
+    public func endBatch() -> Success {
+        // When not batching, just upload whatever is left over
+        guard let token = batchToken else {
+            return uploadPayload(self.currentPayload)
+                >>== effect({ self.ifUnmodifiedSince = $0.value.modified })
+                >>== effect({ self.onCollectionUploaded?($0.value) })
+                >>> succeed
+        }
+
+        return self.commit(token)
+    }
+
+    private func start(payload: Payload) -> Deferred<Maybe<BatchToken?>> {
+        let batchStartParam = NSURLQueryItem(name: "batch", value: "true")
+        let lines = payload.records.map { $0.payload }
+        return self.uploadPayload(payload, queryParams: [batchStartParam]) >>== { storageResponse in
+            self.ifUnmodifiedSince = storageResponse.value.modified
+
+            if let batchToken = storageResponse.value.batchToken {
+                log.debug("Uploaded \(lines.count) records and received batch token \(batchToken)")
+                return deferMaybe(batchToken)
+            }
+
+            // Since we were unable to start a batch, the records that were sent will be committed directly to
+            // the collection so we should update our timestamp and invoke our upload callback.
+            log.debug("Uploaded \(lines.count) records but received no batch token - records will be committed to collection right away")
+            self.ifUnmodifiedSince = storageResponse.value.modified
+            self.onCollectionUploaded?(storageResponse.value)
+            return deferMaybe(nil)
+        }
+    }
+
+    private func commit(token: BatchToken) -> Success {
+        let batchParam = NSURLQueryItem(name: "batch", value: token)
+        let commitParam = NSURLQueryItem(name: "commit", value: "true")
+
+        // We could send a payload when committing the batch but that would involve us slicing up the in-progress 
+        // payload to make it fit just right. Easier to just push it up in a new batch.
+        return self.uploadPayload(Payload(config: self.config), queryParams: [batchParam, commitParam])
```

Comment:
I'm confused. `addRecord` never allows the `Payload` to expand beyond its allowable size. So we never need to slice up a payload; we can always upload it here.

Furthermore, I don't see how an upload happens in the simplest case!

```
let client = Sync15BatchClient<T>(…)
client.addRecord(myRecord)
// 50 times.
client.addRecord(anotherRecord)   // 51st, so there's a token
client.endBatch()
```

… looks like that adds the last record to the new empty payload, calls `commit`, which then sends an empty payload, leaving the final record not uploaded.